sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2109,6 +2110,39 @@ class Analyzer(override val catalogManager: CatalogManager)
     }
   }
 
+  /**
+   * Resolves `UnresolvedAttribute` to `OuterReference` if we are resolving subquery plans (when
+   * `AnalysisContext.get.outerPlan` is set). Note that, this rule must be run after
+   * `ResolveReferences` and `ResolveMissingReferences`, as outer reference should be the last
+   * resort during column resolution.
+   */
+  object ResolveOuterReferences extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = {
+      // Only apply this rule if we are resolving subquery plans.
+      if (AnalysisContext.get.outerPlan.isEmpty) return plan
+
+      val preparedPlan = ResolveMissingReferences(ResolveReferences(plan))
```

Comment:
Is applying these rules once enough? e.g. is it possible that the application of `ResolveMissingReferences` makes more attributes resolvable by `ResolveReferences` and so forth?

Reply:
as long as an `UnresolvedAttribute` can't be resolved by these rules, we can resolve it to outer references.

Reply:
Also these rules won't generate any new `UnresolvedAttribute ` correct?